components/admin_console/admin_settings.tsx

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {Overlay, Tooltip} from 'react-bootstrap';
+
+import {localizeMessage} from 'utils/utils.jsx';
+import SaveButton from 'components/save_button';
+import FormError from 'components/form_error';
+
+import AdminHeader from 'components/widgets/admin_console/admin_header';
+
+type Props = {
+    config?: object;
+    environmentConfig?: object;
+    setNavigationBlocked?: (blocked: boolean) => void;
+    updateConfig?: (config: object) => {data: object; error: ClientErrorPlaceholder};
+}
+
+type State = {
+    saveNeeded: boolean;
+    saving: boolean;
+    serverError: string|null;
+    serverErrorId?: string;
+    errorTooltip: boolean;
+}
+
+type StateKeys = keyof State;
+
+// Placeholder type until ClientError is exported from redux.
+// TODO: remove ClientErrorPlaceholder and change the return type of updateConfig
+type ClientErrorPlaceholder = {
+    message: string;
+    server_error_id: string;
+}
+
+export default abstract class AdminSettings extends React.Component<Props, State> {
+    public constructor(props: Props) {
+        super(props);
+        const stateInit = {
+            saveNeeded: false,
+            saving: false,
+            serverError: null,
+            errorTooltip: false,
+        };
+        if (props.config) {
+            this.state = Object.assign(this.getStateFromConfig(props.config), stateInit);
+        } else {
+            this.state = stateInit;
+        }
+    }
+
+    protected abstract getStateFromConfig(config: object): State;
+
+    protected abstract getConfigFromState(config: object): object;
+
+    protected abstract renderTitle(): React.ReactElement;
+
+    protected abstract renderSettings(): React.ReactElement;
+
+    protected handleSaved?: ((config: object) => React.ReactElement);
+
+    protected canSave?: () => boolean;
+
+    private closeTooltip = () => {
+        this.setState({errorTooltip: false});
+    }
+
+    private openTooltip = (e: React.MouseEvent) => {
+        const elm: HTMLElement|null = e.currentTarget.querySelector('.control-label');
+        if (elm) {
+            const isElipsis = elm.offsetWidth < elm.scrollWidth;
+            this.setState({errorTooltip: isElipsis});
+        }
+    }
+
+    private handleChange = (id: StateKeys, value: any) => {
+        this.setState((prevState) => ({
```

Comment:
You don't need to use this approach to set state. The `setState` function always preserves the old existing state and only update the fields that have provided, so this would be the same as passing directly the object with the `saveNeeded` and `[id]` keys.

Reply:
I do not know the details on why it works like that, but typescript finds a type error if you use the "[key_name]: value" unless you use this work-around. The error in this case would be:

`Argument of type '{ [x: string]: any; saveNeeded: true; }' is not assignable to parameter of type 'State | ((prevState: Readonly<State>, props: Readonly<Props>) => State | Pick<State, "saveNeeded" | "saving" | "serverError" | "serverErrorId" | "errorTooltip"> | null) | Pick<...> | null'.`

If you know any other cleaner work-around, or the reason why it behaves like that, it would be great to know.

Thanks!